### ISSUE TITLE
Decide on name for `derive(SmartPtr)`

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#129104]
We need to resolve the open question on:

- https://github.com/rust-lang/rfcs/pull/3621

...regarding what name to use for the thing formerly known an `#[derive(SmartPtr)]` or `#[derive(SmartPointer)]`.

Tracking:

- https://github.com/rust-lang/rust/issues/123430

**Comment by ChrisDenton** [comment]
I'll say I don't like the term "smart pointer" in Rust. It has a more specific meaning in C++ but porting that terminology over to Rust has been less than smooth. Like the [Deref](https://doc.rust-lang.org/stable/std/ops/trait.Deref.html) docs still imply that anything implementing the `Deref` trait is a "smart pointer" which is a bit of stretch on its own. But now `#[derive(SmartPtr)]` doesn't imply `Deref` and everything gets even more muddy.

I'd rather we use terms we can fully define without confusion. This feels similar to the "object safe" debate in that one aspect.

**Comment by Darksonn** [comment]
One option is to name it in relation to what it enables. For example: `#[derive(DynamicDispatch)]` or `#[derive(DynamicPtr)]`.


@rustbot label F-derive_smart_pointer

**Comment by crlf0710** [comment]
My personal two cents :)
```rust
#[repr(owned_pointer)] // this also implies `#[repr(transparent)]` somehow
struct MyBox<T: ?Sized>(Box<T>);
```

**Comment by Darksonn** [comment]
@crlf0710 Owned pointer is not good because it can be used with pointers that don't have ownership of the value.

I propose to use the name `#[derive(DynamicPtr)]`. Any objections?

**Comment by nikomatsakis** [comment]
> One option is to name it in relation to what it enables

I like this principle...

> For example: #[derive(DynamicDispatch)] or #[derive(DynamicPtr)].

...but neither of these is "self explanatory" to me. What it enables specifically is the possibility of `self: MyType<dyn Bar>` appearing in traits and supporting dynamic dispatch, basically? Is that *it*? I am trying to page this back in.

**Comment by Darksonn** [comment]
It does two things. One, it allows coercions from `MySmartPointer<ConcreteType>` to `MySmartPointer<dyn MyTrait>`.
```rs
fn main() {
    let ptr: MySmartPointer<i32> = MySmartPointer(Box::new(4));

    // This coercion would be an error without the derive.
    let ptr: MySmartPointer<dyn MyTrait> = ptr;
}
```
The other thing is that it makes `self: MySmartPointer<Self>` object-safe in traits.
```rs
trait MyTrait {
    // Arbitrary self types is enough for this.
    fn func(self: MySmartPointer<Self>);
}

// But this requires #[derive(SmartPointer)].
fn call_func(value: MySmartPointer<dyn MyTrait>) {
    value.func();
}
```

**Comment by nikomatsakis** [comment]
After [discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/425075-rust-for-linux/topic/derive.20smart.20pointer/near/474354384), Alice and I would like to propose `derive(DynCapablePointer)`.

**Comment by nikomatsakis** [comment]
I'm going to go ahead and move to merge to move this up the lang priority

@rfcbot fcp merge

**Comment by compiler-errors** [comment]
It's general unsizing, since this macro expands to an implementation of `CoerceUnsized`.

**Comment by nikomatsakis** [comment]
@rfcbot fcp cancel

Canceling FCP because we are live discussing alternatives.

**Comment by traviscross** [comment]
@rfcbot fcp merge

We talked through this in the planning meeting today, and in the end, what Josh, Alice, and I came away happy with (the meeting had ended by the time we landed here) is:

```rust
#[derive(UnsizePointee)]
```

So I propose that for FCP merge.  Read on for why.

We rejected `#[derive(DynReceiver)]` after confirming that this does not in fact derive `Receiver`, which would make that name weird.

We then considered `#[derive(DynCapablePointer)]`.  There are some problems with this.

First, and trivially, we'd earlier leaned toward saying `Ptr` rather than `Pointer` out of consistency with existing things.

Second, we'd just decided to call traits that are compatible with `dyn` "dyn compatible" rather than "dyn capable".  We considered whether this is in fact different, and decided that it isn't really.

That left us with `#[derive(DynCompatiblePtr)]`.  That's just kind of an unpleasantly long name, so we considered shorter alternatives like `#[derive(DynPtr)]` and `#[derive(DynablePtr)]`.  We momentary liked `DynablePtr`.

But there's a meaningful problem with using "dyn" here.  The inner type can also be sliced, and that really has nothing to do with it being "dynable".  So "dyn" here implies the wrong causality.  The true causality is that the things are dynable because they can be unsized, and they can be sliced also because they can be unsized.

On that basis, we considered `#[derive(UnsizablePtr)]`.  Against this was an ambiguity about whether this is "un(sizeable) ptr" or "unsize(able) ptr".  However, we didn't really want to drop back to `UnsizeCompatiblePtr`.  For one thing, there's some precedent that we're setting in this choice.  If we do an `XCompatibleThing`, then it's probably going to make it less likely that we'd ever do `XableThing` instead, and in most cases, if something is "X compatible", then it's going to be "Xable", so it'd be an unfortunate precedent.

Two great observations at this point led to the ultimate solution:

- We're not really talking about the capabilities of the *pointer* here, we're talking about the capabilities with respect to the *pointee*.
- Traits like `Unsize` and `Copy` already imply capabilities -- we don't say `Copyable` or `CopyCompatible`.

On this basis, we settled on...

```rust
#[derive(UnsizePointee)]
```

That is, we're asking to get back a type where we can act to unsize the pointee.  (In terms of how it sounds, consider, comparably, e.g. `derive(CopyPointee)`, if that were a thing).

That's where we landed.  How did we do?  What does the rest of the team think?

**Comment by joshtriplett** [comment]
Fortunately, we reached a consensus, and did not have to fall back to `PointeeMcPointface`. We, ahem, `yeet`ed that suggestion once we had a better one. 😁

**Comment by traviscross** [comment]
@rfcbot cancel

Wait, no, sorry, we thought of a slightly better idea.

**Comment by traviscross** [comment]
@rfcbot fcp merge

Final answer, in later discussion with Josh:

```rust
#[derive(UnsizeInner)]
```

Incorporate by reference everything said [above](https://github.com/rust-lang/rust/issues/129104#issuecomment-2389807803) (which I'll also edit), and add that:

We use "inner" a lot in the standard library... we don't use "pointee" a lot anywhere.  From the user's perspective, this derive allows going from e.g. `Thing<T>` to `Thing<dyn T>`.  In Rust, we'd say that's the "inner" type.  So let's call it that.

**Comment by lqd** [comment]
Would the associated `#[pointee]` attribute be changed to `#[inner]`?

**Comment by traviscross** [comment]
That would probably make sense, for both consistency and for the same reasons.

**Comment by tmandry** [comment]
I don't know about "inner". Unsizing is a transformation that affects the memory of the pointer, not on the value it points to. It's weird to talk about because it also happens on the type parameter of the pointer. But in every case where we use "inner" in the standard library, it's referring to the value and not just the type parameter.

I agree that the -able suffix would be a departure. If we were going in the direction of using "unsize" as a verb, I think `UnsizePointee` or `UnsizeReferent` both work better.

Now, given a choice between a name that starts with "un" and another name, I'm almost always going to choose the other name. Someone made a point in the meeting (and I agree) of preferring to use "positive" names when possible. This also would exacerbate the problem of `Unpin`, which also has to do with pointers and is quite confusing to users (especially when talking about a type that is `!Unpin`). Adding another "un" name to the mix is not helping matters.

So are there any other operations we name it after? There are a few operations we could say are enabled by this derive:

* Unsizing coercion
* Dynamic dispatch
* Accessing the pointer metadata

The last one, accessing pointer metadata, is required for both unsizing coercion and dynamic dispatch. In fact, whatever we name this trait could a bound we would use on a generalized version of the (unstable) [`ptr_metadata`](https://doc.rust-lang.org/stable/std/intrinsics/fn.ptr_metadata.html) intrinsic. Currently its signature hardcodes raw pointers, but it could be adapted easily.

```rust
// Signature today
pub const fn ptr_metadata<T, M>(ptr: *const T) -> M
where
    P: Pointee<Metadata = M> + ?Sized;

// Generalized signature
pub const fn ptr_metadata<P, T, M>(ptr: P) -> M
where
    P: PtrMetadata<Pointee = T>,
    T: Pointee<Metadata = M> + ?Sized;
```

Operationally, this captures one side of what the complier needs to do. The other is to be able to construct a pointer given an unsized type and its metadata. But both of these are about metadata, and one might say that `PtrMetadata` captures them both nicely enough while still leaving the "nicest" names for later. I'd like to hear what other members of the team think about this.

Or if we want to go further and describe the essence of what a metadata-carrying type with coercion might be, we could just call it `Pointer`.

**Comment by RalfJung** [comment]
> We say "inner" a lot in Rust, e.g. in the standard library. We don't often say "pointee"

This part of the comment is now quite confusing, since "pointee" was not previously mentioned in the post at all. I had to go through the edit history of that post to see that previously the proposal was `UnsizePointee`, and only then did the current text make sense to me.

FWIW I have a slight preference for `UnsizePointee` over `UnsizeInner`, since the latter seems rather non-specific. This is about a pointer type after all. But I assume it will usually be wrapping another pointer type, so `#[pointee]` is an odd name for the attribute -- shouldn't it be `#[pointer]`?

**Comment by traviscross** [comment]
The history is in a footnote above.

**Comment by Darksonn** [comment]
When I left the meeting we had settled on `UnsizePointee`, but it sounds like things continued to `UnsizeInner`. One challenge with inner is that the derive only applies to smart pointers and not [transparent containers](https://rust-lang.github.io/rfcs/3621-derive-smart-pointer.html#transparent-containers). The name `Inner` sounds general enough that you'd want it to work with transparent containers too, but the derive doesn't and can't.

As for `Pointer` vs `Pointee`, yes, the thing whose bytes in memory actually changes is the pointer, but on the type-system level what's happening is that the pointee gets unsized according to the trait called `Unsize`.

**Comment by RalfJung** [comment]
> The history is in a footnote above.

The main text should be understandable without reading the footnote, and right now, the term "pointee" falls from the sky here at the end of the post.

**Comment by traviscross** [comment]
Sure, I've added...

> *(Please see the footnote [^1] for some description of the history of proposals.)*

...to the body.

**Comment by traviscross** [comment]
@tmandry:

You raise an interesting point about the metadata angle.  Something to think about.

About this:

> I don't know about "inner". Unsizing is a transformation that affects the memory of the pointer, not on the value it points to. It's weird to talk about because it also happens on the type parameter of the pointer. But in every case where we use "inner" in the standard library, it's referring to the value and not just the type parameter.

Yes, I had a roughly similar thought, but decided it was OK anyway because it does feel more natural to me to focus on the type.  Though when you say...

> If we were going in the direction of using "unsize" as a verb, I think `UnsizePointee` or `UnsizeReferent` both work better.

...it seems that this particular argument would apply equally to "pointee" as much as to "inner".  Perhaps that is what you meant?

The framing that led us down this path after the meeting was that the `Unsize` capability applied to the inner type, or pointee, or whatnot, rather than the pointer, which is how we ended up with `UnsizePointee` initially.

So given, e.g., [`Arc<T>::into_inner`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_inner) that returns (an option of) that inner `T`, at a type level, it feels like we'd also say the `T` is the "inner" thing and that's what we're unsizing.

Also, it just feels a bit to me that we mostly treat the "pointerness" of things a bit abstractly, maybe not quite as an implementation detail, but close.  An `Arc` is of course a "smart pointer", but we didn't name it `ArcPtr`, and we don't typically focus on its pointerness when explaining it.  Again, I think it is the focus on types that allows us to look at it that way.

But, hey, I can see it the other way too, so I'm also interested to hear what people think when we take this back up on a call -- we'll leave it nominated.

**Comment by traviscross** [comment]
> But I assume it will usually be wrapping another pointer type, so `#[pointee]` is an odd name for the attribute -- shouldn't it be `#[pointer]`?

This is actually one of the (unstated) reasons I preferred "inner", was to not have to tackle this very question.

> One challenge with inner is that the derive only applies to smart pointers and not [transparent containers](https://rust-lang.github.io/rfcs/3621-derive-smart-pointer.html#transparent-containers). The name `Inner` sounds general enough that you'd want it to work with transparent containers too, but the derive doesn't and can't.

You're right of course, and this particular bit was and remains my most nagging concern about "inner".  Agree this is a challenge and is a plausible reason to want something else.  It might be useful to think comparatively about what we would want to name a comparable derive for such transparent containers.

**Comment by Darksonn** [comment]
> FWIW I have a slight preference for `UnsizePointee` over `UnsizeInner`, since the latter seems rather non-specific. This is about a pointer type after all. But I assume it will usually be wrapping another pointer type, so `#[pointee]` is an odd name for the attribute -- shouldn't it be `#[pointer]`?

The attribute is used like this:
```rs
#[derive(SmartPointer)]
#[repr(transparent)]
struct MySmartPointer<#[pointee] T: ?Sized, U> {
    ptr: Box<T>,
    _phantom: PhantomData<U>,
}
```
So yes, the custom smart pointer is going to wrap another pointer type, in this case `Box<T>`. But the `#[pointee]` attribute is annotated on the pointee type `T`, not on the pointer type `Box<T>`.

**Comment by Darksonn** [comment]
At that point you should just keep the original name. The argument against `SmartPointer` is that although the macro only applies to smart pointers, it doesn't apply to *all* smart pointers.

**Comment by traviscross** [comment]
Remind me, if you would, this is about smart pointers that are not transparent?  The RFC and some of the thread discusses the other direction, that deriving this by itself doesn't make a type a smart pointer.

**Comment by Darksonn** [comment]
Transparent isn't an important factor here. Right now, we require `#[repr(transparent)]` to use the macro, but that's not a particularly important detail. It's required by the specific way that unsizing coercions are implemented in the compiler right now, but there's no fundamental reason it's required.

The reason that the derive does not apply to all smart pointers is that a smart pointer does not _necessarily_ support `T: ?Sized` types, and this derive is specifically about adding `T: ?Sized` support to a pointer type. For example, the kernel's `ARef` type has a `T: AlwaysRefcounted` bound and I don't think the `AlwaysRefcounted` trait is object safe. However, `ARef` still qualifies as a smart pointer imo. It's basically an `Arc`.

> that deriving this by itself doesn't make a type a smart pointer.

It doesn't make something a smart pointer, but it can only be used with things that are already smart pointers.

Due to the way unsizing coercions work, the derive can _only_ be applied to structs where the generic type parameter annotated with `#[pointee]` is stored behind a single layer of indirection:

* When the generic parameter is stored behind zero layers of indirection, you are dealing with a transparent container [which I discussed earlier](https://github.com/rust-lang/rust/issues/129104#issuecomment-2390615027). The macro would need to generate different where bounds for those, and there's no way to tell purely syntactically which case you are dealing with.
* When the generic parameter is stored behind two or more layers of indirection, then unsizing coercions are impossible because they work by replacing a thin pointer with a wide pointer, which we can't do in this case. There's no way convert an `Arc<Box<MyStruct>>` to `Arc<Box<dyn MyTrait>>` because that requires changing the contents of the allocation referenced by the `Arc`.

So I do think it's fair to say that it can only be used with things that can be called smart pointers.

**Comment by traviscross** [comment]
Thanks for that great analysis.  So, to apply this to @tmandry's framing, which is that access to the metadata is the key thing here enabling both unsizing and dynamic dispatch, we would say that, even if we had a `Pointer` trait for such access to the metadata, that this derive would not achieve providing that to all types that we would reasonably consider to be smart pointers.

So we would still need to come up with a name other than `derive(Pointer)` that would somehow encode these limitations.

**Comment by traviscross** [comment]
@rfcbot fcp cancel

I've growing cold on `derive(UnsizeInner)` due to transparent containers.  Maybe it'd actually be a better name for that.

I'm probably warming back up to `derive(UnsizePointee)`, but let's propose FCP again after the discussion on this thread plays out and we all talk it through on a call.

(With apologies to @Darksonn whom I know wants an answer here.)

**Comment by nikomatsakis** [comment]
This is kind of amusing =)

I think the reason it's so hard for us to settle on a name is that this derive is not a "complete" abstraction -- it doesn't create a workable smart pointer, it extends a pointer (and not necessarily a "smart" one)[^smart] so that it can accommodate unsized referents/pointees.

I've said a few times that I would rather that this derive is "hidden" a bit, e.g., `std::marker` or some such place, because I don't really think it's our final solution, just an interim stop.

It doesn't quite fit our naming convention of "actions" (e.g., `Send`, `Copy`), but `PtrToUnsized` might work, or `PtrToDyn`, or `DynCompatiblePtr`. I think the challenge here is that we don't have a name for the "action" of "coercing to a wide pointer", nor for that matter a standard name for "wide pointer" (that is, pointer to some `T: !Sized`).

If we do want to go with action, I suppose the *action* here is coercion...maybe `CoerceToWidePtr`? `CoerceToUnsizePtr`?

To be honest I think we're overthinking this and we should run with `#[derive(std::marker::DynCompatiblePtr)]` and just accept that it doesn't fit our naming conventions because it's a weird special case thing. Putting it in marker makes it feel better to me because it's a kind of marker. 

[^smart]: What the heck is a smart pointer anyway? My take is that it is something that "acts like a reference", i.e., gives transparent access to its referenced value, but which adds some add'l behavior. Hence `Rc` and `Arc` are "smart pointers", as is `&T` but in a trivial way. `*const T` is *not* a smart pointer. But note that this derive is useful for `*const T` as well.

**Comment by nikomatsakis** [comment]
Regarding some of the names that have been proposed so far...

* I don't care for `Inner` because it's too generic.
* I prefer `derive(Pointer)` to `derive(SmartPointer)`. It's *awfully* general, and I suspect we may wish we had kept that name for something else, but also, I don't know what it means to be a "pointer" in some general sense but I guess maybe the only thing you can say about pointers that holds in general is that they are *indirect*... and hence they can refer to things without a known size.
* The `CoerceToX` family of names feels "ok", except that I don't really know what `X` should be. I don't know how I feel about committing to the name coercion but it seems like a pretty widely used term.
* The `DynCompatible` prefix feels "ok", even though it doesn't describe the fact that it's *also* compatible with slices. I could also live with `UnsizedCompatiblePtr`, or perhaps `DynSizedCompatible`.

Some more musings since I am apparently effectively nerdswiped:

* Musing on that last point a bit, I suppose the "positive" form of "Unsized Compatible" is that it supports *arbitrary* pointees? `#[derive(AnyPointee)]`?

**Comment by traviscross** [comment]
`CoercePointee` sounds right to me and like the high water mark here.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose the following name:

* `#[derive(std::marker::CoercePointee)]`

Reasoning:

* It is named after the primary *action* it enables -- it becomes possible to coerce the pointer's pointee. The other effect, being able to use the type as a `self` type with `dyn Trait` (presuming the type implements `Receiver`) is more a consequence of this.
* The name `Pointee` echoes the (unstable) name of the [`Pointee`](https://doc.rust-lang.org/std/ptr/trait.Pointee.html) trait. [The `Pointee` RFC actually does not justify the choice `Pointee` was chosen](https://github.com/rust-lang/rfcs/blob/master/text/2580-ptr-meta.md#unresolved-questions), but I presume it is because `Referent` confusable with reference. Perhaps @SimonSapin or somebody from @rust-lang/libs-api would care to weigh in on this point.
* It avoids baking in jargon-y terms like unsize, wide pointer, or smart pointer.

**Comment by tmandry** [comment]
This is better. I prefer "Referent" over "Pointee" because of the precedent it sets and because "Referent" feels more approachable. But I think I can live with either one.

@rfcbot reviewed

(I also argue that "wide pointer" is not more jargon-y than "coerce", but in any case I agree that naming the derive after an action is preferable.)

**Comment by nikomatsakis** [comment]
@tmandry I'm open to `CoerceReferent`. I too think it sounds better.

**Comment by nikomatsakis** [comment]
@rfcbot concern pointee-or-referent

I'm filing a concern to finalize the naming convention here.

I'd be curious to get a "straw poll" of whether people prefer `Pointee` (react with 🎉) or `Referent` (react with ❤).

**Comment by SimonSapin** [comment]
`Pointee` was the only thing I could come up with for "the thing behind a pointer". I didn’t think of `Referent` at the time. So please don’t take the original RFC as having a strong position on naming and feel free to rename if something else is better.

**Comment by nikomatsakis** [comment]
@rfcbot resolve pointee-or-referent

I switched the text to `CoerceReferent`

**Comment by traviscross** [comment]
@rfcbot reviewed

Great outcome.  Naming is hard.

Let's CC the RFC 2580 tracking issue...

- https://github.com/rust-lang/rust/issues/81513

...and we'll mention this there, since this FCP will affect the naming of those APIs.

**Comment by dingxiangfei2009** [comment]
@nikomatsakis By extension, I suppose that we wanted to switch from `#[pointee]` to `#[referent]` as well.

**Comment by nikomatsakis** [comment]
@joshtriplett writes on Zulip:

> I personally prefer Pointee as being unambiguously "what a pointer points to", rather than having one more thing with "ref" in it, and this one not being about references.

While I think "referent" is a more elegant name, I don't feel super strongly about this. Looking at the (limited) votes on [this comment](https://github.com/rust-lang/rust/issues/129104#issuecomment-2392292010), I see that `Pointee` and `Referent` are currently tied. 

@rust-lang/lang or @rust-lang/libs, somebody want to weigh in? (I recommend you leave an emoji on [this comment](https://github.com/rust-lang/rust/issues/129104#issuecomment-2392292010).)

**Comment by nikomatsakis** [comment]
OK, the people have spoken lol. I'm going to flip back to `Pointee`.

![Image](https://github.com/user-attachments/assets/03436d2d-9466-4e4d-8b4c-20f6db1555f0)

**Comment by nikomatsakis** [comment]
@rfcbot resolve pointee-or-referent

**Comment by traviscross** [comment]
To reiterate, give the back and forth, what we're FCPing here is the decision to go with:

```rust
#[derive(CoercePointee)]
```

**Comment by kupiakos** [comment]
My only issue with `CoercePointee` is that this is coercion of a pointee, but this derive only applies to the unsizing kind:

```rust
struct Foo<T>(Box<T>);
fn coerce_pointee<'a>(x: Foo<&'static i32>, _: &'a i32) -> Foo<&'a i32> {
    x
}
```

Personally I prefer `UnsizePointee` as it describes the specific type of coercion it enables.

**Comment by tmandry** [comment]
That's true. I think I'm okay with it because we still allow the other kinds of coercion that make sense for a pointer (your code sample already compiles, in other words). I can't think of another kind of coercion we would want to allow behind a pointer type that isn't either already allowed or covered by this derive.

On a related note, I don't hate the idea of sticking with the already established `CoerceUnsized`, but I do think _both_ `CoercePointee` and `UnsizePointee` are better and clearer, and I still prefer the first (which are are FCP'ing).

**Comment by Darksonn** [comment]
What name did we end up with?

**Comment by lqd** [comment]
`CoercePointee`, cf https://github.com/rust-lang/rust/issues/129104#issuecomment-2403176941

**Comment by tmandry** [comment]
I think it would still be required, I don't recall any discussion about the path and my understanding was that this would _not_ be in the standard prelude to begin with.

**Comment by Darksonn** [comment]
The intent is that this will not be in the prelude.

**Comment by Darksonn** [comment]
Should this be closed now that #131284 is merged?
